dev/cluster.go

```diff
@@ -0,0 +1,288 @@
+package dev
+
+import (
+	"bytes"
+	"context"
+	goerrors "errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+
+	"github.com/go-logr/logr"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var defaultSchemeBuilder runtime.SchemeBuilder = runtime.SchemeBuilder{
+	clientgoscheme.AddToScheme,
+	apiextensionsv1.AddToScheme,
+}
+
+type ClusterConfig struct {
+	Logger        logr.Logger
+	SchemeBuilder runtime.SchemeBuilder
+	NewWaiter     ClusterNewWaiterFunc
+	WaitOptions   []WaitOption
+}
+
+type ClusterNewWaiterFunc func(
+	client client.Client, scheme *runtime.Scheme,
+	defaultOpts ...WaitOption,
+) *Waiter
+
+func (c *ClusterConfig) Default() {
+	if c.Logger.GetSink() == nil {
+		c.Logger = logr.Discard()
+	}
+	if c.NewWaiter == nil {
+		c.NewWaiter = NewWaiter
+	}
+	if c.WaitOptions == nil {
+		c.WaitOptions = append(c.WaitOptions, WithLogger(c.Logger))
+	}
+}
+
+type ClusterOption interface {
+	ApplyToClusterConfig(c *ClusterConfig)
+}
+
+// Container object to hold kubernetes client interfaces and configuration.
+type Cluster struct {
+	Scheme     *runtime.Scheme
+	RestConfig *rest.Config
+	CtrlClient client.Client
+	Waiter     *Waiter
+	WorkDir    string
+	// Path to the kubeconfig of the cluster
+	Kubeconfig string
+
+	config ClusterConfig
+}
+
+// Creates a new Cluster object to interact with a Kubernetes cluster.
+func NewCluster(workDir string, opts ...ClusterOption) (*Cluster, error) {
```

Comment:
nit: feels every comment should be a helper function:

- `addDefaultSchemes`
- `applyOptions`
- `applySchemesFromOptions`
- etc.

Reply:
Using this as thread for all "Function too long"/"Refactor into helper functions" comments.

Helper functions:
I try to not introduce extra indirection until needed, which is usually because it either helps in (unit-)testing or to keep the code DRY. Spreading the code over many small functions often makes the code harder to refactor and to understand function names that are obvious to the author are often less clear to other developers (or become unwieldy long :D)

IMHO it's very easy to fall into a trap of premature optimization with helper functions. I have spend plenty time going through multiple layers of helper functions (that are not used anywhere else)  and having to either refactor all of them to add a parameter (ctx for example :grimacing: ) or rewrite them from scratch to support some new code branch.
In both cases the effort of putting everything into a lot of small function hindered refactoring way more than it benefited.

From my experiences code bases become more DRY, helper functions more frequent as the software matures and no large refactoring or big changes are needed. Most Projects never get so far, because they are superseded entirely way before they get so far.

Long functions:
In my book a long function starts somewhere around ~100 LoC.
I think it's way more important to keep complexity within a function low.
If you do early returns, keep branches and loops to a minimum (or move branch/loop logic into new functions) even 200 LoC functions can be like comprehensive shopping list.

And if your shopping list is 100 items long, splitting it over 20 post-it notes may not be the improvement you are looking for :D

All of this is of cause very much down to personal preference and the trauma of previous code bases I have worked on...

Reply:
Go's a tricky one. The issue was never really length, but scope creep. And Go handled this deftly with `if` scoped assignments. If you actually look at the namespace of `NewCluster` there's at most four names in use. The comments could also be removed here without loss of understanding. Now something like `HelmInstall` is smellier because `HelmCmd` is probably it's own type that could be initialized from a cluster which would reduce repeating initialization in `execHelmCommand` and passing around parameters. But it's really situational so don't think there's any formula to apply.

Reply:
Oh HelmInstall is smelly for a bunch of reasons. I have to admit that I have hacked it together in the hope I can get rid of it again :)

You are right Helm should be moved into it's own struct especially if we want to expose more of helms functionality.